File: src/BehaviorsSDKNative/Microsoft.Xaml.Interactions/IncrementalUpdateBehavior.cpp
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.
#pragma warning(disable: 6298)
#include "pch.h"
#include "IncrementalUpdateBehavior.h"
#include "ResourceHelper.h"

using namespace std;
using namespace Platform;
using namespace Windows::Foundation;
using namespace Windows::UI::Xaml;
using namespace Windows::UI::Xaml::Controls;
using namespace Windows::UI::Xaml::Controls::Primitives;
using namespace Windows::UI::Xaml::Media;
using namespace Microsoft::Xaml::Interactivity;
using namespace Microsoft::Xaml::Interactions::Core;

namespace DependencyProperties
{
	static DependencyProperty^ Phase = DependencyProperty::Register(
		"Phase",
		int::typeid,
		IncrementalUpdateBehavior::typeid,
		ref new PropertyMetadata(1, ref new PropertyChangedCallback(&IncrementalUpdateBehavior::OnPhaseChanged)));

	static DependencyProperty^ IncrementalUpdater = DependencyProperty::RegisterAttached(
		"IncrementalUpdater",
		IncrementalUpdateBehavior::IncrementalUpdater::typeid,
		IncrementalUpdateBehavior::typeid,
		ref new PropertyMetadata(nullptr, ref new PropertyChangedCallback(&IncrementalUpdateBehavior::OnIncrementalUpdaterChanged)));
}

void IncrementalUpdateBehavior::IncrementalUpdater::PhasedElementRecord::FreezeAndHide()
{
	if (_isFrozen)
	{
		return;
	}

	_isFrozen = true;
	_localOpacity = _frameworkElement->ReadLocalValue(FrameworkElement::OpacityProperty);
	_localDataContext = _frameworkElement->ReadLocalValue(FrameworkElement::DataContextProperty);
	_frameworkElement->Opacity = 0.0;
	_frameworkElement->DataContext = _frameworkElement->DataContext;
}

void IncrementalUpdateBehavior::IncrementalUpdater::PhasedElementRecord::ThawAndShow()
{
	if (!_isFrozen)
	{
		return;
	}

	if (_localOpacity != DependencyProperty::UnsetValue)
	{
		_frameworkElement->SetValue(FrameworkElement::OpacityProperty, _localOpacity);
	}
	else
	{
		_frameworkElement->ClearValue(FrameworkElement::OpacityProperty);
	}

	if (_localDataContext != DependencyProperty::UnsetValue)
	{
		_frameworkElement->SetValue(FrameworkElement::DataContextProperty, _localDataContext);
	}
	else
	{
		_frameworkElement->ClearValue(FrameworkElement::DataContextProperty);
	}

	_isFrozen = false;
}

UIElement^ IncrementalUpdateBehavior::IncrementalUpdater::FindContentTemplateRoot(FrameworkElement^ phaseElement)
{
	DependencyObject^ ancestor = static_cast<DependencyObject^>(phaseElement) ;
	while (ancestor != nullptr)
	{
		DependencyObject^ parent = VisualTreeHelper::GetParent(ancestor);
		SelectorItem^ item = dynamic_cast<SelectorItem^>(parent);

		if (item != nullptr)
		{
			return item->ContentTemplateRoot;
		}
		ancestor = parent;
	}

	return nullptr;
}

void IncrementalUpdateBehavior::IncrementalUpdater::CachePhaseElement(FrameworkElement^ phaseElement, int phase)
{
	if (phase < 0)
	{
		throw ref new OutOfBoundsException("phase");
	}

	if (phase <= 0)
	{
		return;
	}

	UIElement^ contentTemplateRoot = IncrementalUpdateBehavior::IncrementalUpdater::FindContentTemplateRoot(phaseElement);
	if (contentTemplateRoot != nullptr)
	{
		// get the cache for this element
		auto& elementCacheRecord = _elementCache[contentTemplateRoot];

		// get the cache for this phase
		auto& elementPhaseRecord = elementCacheRecord[phase];

		// insert the element
		PhasedElementRecord phasedElementRecord(phaseElement);
		elementPhaseRecord.push_back(phasedElementRecord);
	}
}

void IncrementalUpdateBehavior::IncrementalUpdater::UncachePhaseElement(FrameworkElement^ phaseElement, int phase)
{
	if (phase <= 0)
	{
		return;
	}

	UIElement^ contentTemplateRoot = IncrementalUpdateBehavior::IncrementalUpdater::FindContentTemplateRoot(phaseElement);
	if (contentTemplateRoot != nullptr)
	{
		// get the cache for this element
		auto& elementCacheRecord = _elementCache[contentTemplateRoot];

		// get the cache for this phase
		auto& elementPhaseRecord = elementCacheRecord[phase];

		// remove the element
		for (auto it = elementPhaseRecord.begin(); it < elementPhaseRecord.end(); it++)
		{
			if (it->_frameworkElement == phaseElement)
			{
				elementPhaseRecord.erase(it);
				break;
			}
		}
	}
}

void IncrementalUpdateBehavior::IncrementalUpdater::OnContainerContentChanging(Windows::UI::Xaml::Controls::ListViewBase^ sender, ContainerContentChangingEventArgs^ e)
{
	UIElement^ contentTemplateRoot = e->ItemContainer->ContentTemplateRoot;

	if (_elementCache.find(contentTemplateRoot) == _elementCache.end())
	{
		return;
	}

	auto& elementCacheRecord = _elementCache[contentTemplateRoot];

	if (!e->InRecycleQueue)
	{
		for (auto& phaseRecordPair : elementCacheRecord)
		{
			for (auto& phaseElementRecord : phaseRecordPair.second)
			{
				phaseElementRecord.FreezeAndHide();
			}
		}

		if (elementCacheRecord.size() > 0)
		{
			e->RegisterUpdateCallback(elementCacheRecord.cbegin()->first, _cccCallbackHandler);
		}

		// set the DataContext manually since we inhibit default operation by setting e.Handled=true
		FrameworkElement^ frameworkElement = dynamic_cast<FrameworkElement^>(contentTemplateRoot);
		if (frameworkElement != nullptr)
		{
			frameworkElement->DataContext = e->Item;
		}
	}
	else
	{
		// clear the DataContext manually since we inhibit default operation by setting e.Handled=true
		contentTemplateRoot->ClearValue(FrameworkElement::DataContextProperty);

		for (auto& phaseRecordPair : elementCacheRecord)
		{
			for (auto& phaseElementRecord : phaseRecordPair.second)
			{
				phaseElementRecord.ThawAndShow();
			}
		}
	}

	e->Handled = true;
}

void IncrementalUpdateBehavior::IncrementalUpdater::OnContainerContentChangingCallback(Windows::UI::Xaml::Controls::ListViewBase^ sender, ContainerContentChangingEventArgs^ e)
{
	map<int, vector<PhasedElementRecord>> *elementCacheRecord = &_elementCache[e->ItemContainer->ContentTemplateRoot];

	auto it = elementCacheRecord->find(e->Phase);

	for (auto& phaseElementRecord : it->second)
	{
		phaseElementRecord.ThawAndShow();
	}

	it++;
	if (it != elementCacheRecord->end())
	{
		e->RegisterUpdateCallback(it->first, _cccCallbackHandler);
	}
}

void IncrementalUpdateBehavior::IncrementalUpdater::Attach(DependencyObject^ dependencyObject)
{
	_listViewBase = safe_cast<ListViewBase^>(dependencyObject);

	if (_cccHandler == nullptr)
	{
		_cccHandler = ref new TypedEventHandler<ListViewBase^, ContainerContentChangingEventArgs^>(this, &IncrementalUpdateBehavior::IncrementalUpdater::OnContainerContentChanging);
	}

	if (_cccCallbackHandler == nullptr)
	{
		_cccCallbackHandler = ref new TypedEventHandler<ListViewBase^, ContainerContentChangingEventArgs^>(this, &IncrementalUpdateBehavior::IncrementalUpdater::OnContainerContentChangingCallback);
	}

	_containerContentChangingToken = _listViewBase->ContainerContentChanging += _cccHandler;
}

void IncrementalUpdateBehavior::IncrementalUpdater::Detach()
{
	_listViewBase->ContainerContentChanging -= _containerContentChangingToken;
	_listViewBase = nullptr;
}

void IncrementalUpdateBehavior::OnPhaseChanged(DependencyObject^ sender, DependencyPropertyChangedEventArgs^ args)
{
	IncrementalUpdateBehavior^ behavior = safe_cast<IncrementalUpdateBehavior^>(sender);
	IncrementalUpdater^ incrementalUpdater = behavior->FindUpdater();
	FrameworkElement^ frameworkElement = dynamic_cast<FrameworkElement^>(behavior->_associatedObject);

	if (incrementalUpdater != nullptr && frameworkElement != nullptr)
	{
		incrementalUpdater->UncachePhaseElement(frameworkElement, safe_cast<int>(args->OldValue));
		incrementalUpdater->CachePhaseElement(frameworkElement, safe_cast<int>(args->NewValue));
	}
}

DependencyProperty^ IncrementalUpdateBehavior::PhaseProperty::get()
{
	return DependencyProperties::Phase;
}

void IncrementalUpdateBehavior::OnIncrementalUpdaterChanged(DependencyObject^ sender, DependencyPropertyChangedEventArgs^ args)
{
	if (args->OldValue != nullptr)
	{
		IncrementalUpdateBehavior::IncrementalUpdater^ incrementalUpdater = static_cast<IncrementalUpdateBehavior::IncrementalUpdater^>(args->OldValue);
		incrementalUpdater->Detach();
	}
	if (args->NewValue != nullptr)
	{
		IncrementalUpdateBehavior::IncrementalUpdater^ incrementalUpdater = static_cast<IncrementalUpdateBehavior::IncrementalUpdater^>(args->NewValue);
		incrementalUpdater->Attach(sender);
	}
}

DependencyProperty^ IncrementalUpdateBehavior::IncrementalUpdaterProperty::get()
{
	return DependencyProperties::IncrementalUpdater;
}

void IncrementalUpdateBehavior::OnAssociatedObjectLoaded(Platform::Object^ sender, Windows::UI::Xaml::RoutedEventArgs^ e)
{
	IncrementalUpdater^ incrementalUpdater = this->FindUpdater();
	FrameworkElement^ frameworkElement = static_cast<FrameworkElement^>(_associatedObject);

	if (incrementalUpdater != nullptr && frameworkElement != nullptr)
	{
		incrementalUpdater->CachePhaseElement(frameworkElement, this->Phase);
	}
}

void IncrementalUpdateBehavior::OnAssociatedObjectUnloaded(Platform::Object^ sender, Windows::UI::Xaml::RoutedEventArgs^ e)
{
	FrameworkElement^ frameworkElement = static_cast<FrameworkElement^>(_associatedObject);

	if (_updater != nullptr && frameworkElement != nullptr)
	{
		_updater->UncachePhaseElement(frameworkElement, this->Phase);
	}

	_updater = nullptr;
}

IncrementalUpdateBehavior::IncrementalUpdater^ IncrementalUpdateBehavior::FindUpdater()
{
	if (_updater != nullptr)
	{
		return _updater;
	}

	DependencyObject^ ancestor = _associatedObject;
	while (ancestor != nullptr)
	{
		DependencyObject^ parent = VisualTreeHelper::GetParent(ancestor);
		ListViewBase^ listView = dynamic_cast<ListViewBase^>(parent);

		if (listView != nullptr)
		{
			IncrementalUpdater^ currentUpdater = IncrementalUpdateBehavior::GetIncrementalUpdater(listView);
			if (currentUpdater == nullptr)
			{
				currentUpdater = ref new IncrementalUpdater();

				IncrementalUpdateBehavior::SetIncrementalUpdater(listView, currentUpdater);
			}
			return currentUpdater;
		}
		ancestor = parent;
	}

	return nullptr;
}

void IncrementalUpdateBehavior::Attach(DependencyObject^ associatedObject)
{
	if (associatedObject == _associatedObject)
	{
		return;
	}

	if (Windows::ApplicationModel::DesignMode::DesignModeEnabled)
	{
		return;
	}

	if (_associatedObject != nullptr)
	{
		throw ref new FailureException(ResourceHelper::GetString("CannotAttachBehaviorMultipleTimesExceptionMessage"));
	}

	_associatedObject = associatedObject;

	FrameworkElement^ frameworkElement = dynamic_cast<FrameworkElement^>(associatedObject);
	if (frameworkElement != nullptr)
	{
		_loadedToken = frameworkElement->Loaded += ref new Windows::UI::Xaml::RoutedEventHandler(this, &IncrementalUpdateBehavior::OnAssociatedObjectLoaded);
		_unloadedToken = frameworkElement->Unloaded += ref new Windows::UI::Xaml::RoutedEventHandler(this, &IncrementalUpdateBehavior::OnAssociatedObjectUnloaded);
	}
}

void IncrementalUpdateBehavior::Detach()
{
	FrameworkElement^ frameworkElement = dynamic_cast<FrameworkElement^>(_associatedObject);
	if (frameworkElement != nullptr)
	{
		frameworkElement->Loaded -= _loadedToken;
		frameworkElement->Unloaded -= _unloadedToken;
		// no need to perform the work that Unloaded would have done - that's just housekeeping on the cache, which is now going away
	}

	_associatedObject = nullptr;
}

File: src/BehaviorsSDKManaged/Microsoft.Xaml.Interactions/Core/IncrementalUpdateBehavior.cs
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.
namespace Microsoft.Xaml.Interactions.Core
{
    using System;
    using System.Collections.Generic;
    using Windows.UI.Xaml;
    using Windows.UI.Xaml.Controls;
    using Windows.UI.Xaml.Controls.Primitives;
    using Windows.UI.Xaml.Media;
    using Interactivity;

    /// <summary>
    /// A behavior that allows incremental updating of <seealso cref="Windows.UI.Xaml.Controls.ListView"/> and <seealso cref="Windows.UI.Xaml.Controls.GridView"/> contents to support faster updating.
    /// By attaching this behavior to elements in the <seealso cref="Windows.UI.Xaml.Controls.ItemsControl.ItemTemplate"/> used by these views, some of the updates can be deferred until there is render time available, resulting in a smoother experience.
    /// </summary>
    public sealed class IncrementalUpdateBehavior : Behavior<FrameworkElement>
    {
        /// <summary>
        /// Identifies the <seealso cref="Phase"/> dependency property.
        /// </summary>
        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Security", "CA2104:DoNotDeclareReadOnlyMutableReferenceTypes")]
        public static readonly DependencyProperty PhaseProperty = DependencyProperty.Register(
            "Phase",
            typeof(int),
            typeof(IncrementalUpdateBehavior),
            new PropertyMetadata((int)1, new PropertyChangedCallback(IncrementalUpdateBehavior.OnPhaseChanged)));

        /// <summary>
        /// Identifies the <seealso cref="IncrementalUpdater"/> dependency property.
        /// </summary>
        private static readonly DependencyProperty IncrementalUpdaterProperty = DependencyProperty.RegisterAttached(
            "IncrementalUpdater",
            typeof(IncrementalUpdater),
            typeof(IncrementalUpdateBehavior),
            new PropertyMetadata(null, new PropertyChangedCallback(IncrementalUpdateBehavior.OnIncrementalUpdaterChanged)));

        private IncrementalUpdater _updater = null;

        /// <summary>
        /// Gets or sets the relative priority of this incremental update. Lower Phase values are addressed first.
        /// </summary>
        public int Phase
        {
            get { return (int)this.GetValue(IncrementalUpdateBehavior.PhaseProperty); }
            set { this.SetValue(IncrementalUpdateBehavior.PhaseProperty, value); }
        }

        private static void OnPhaseChanged(DependencyObject sender, DependencyPropertyChangedEventArgs args)
        {
            IncrementalUpdateBehavior behavior = (IncrementalUpdateBehavior)sender;
            IncrementalUpdater incrementalUpdater = behavior.FindUpdater();
            FrameworkElement frameworkElement = behavior.AssociatedObject;

            if (incrementalUpdater != null && frameworkElement != null)
            {
                incrementalUpdater.UncachePhaseElement(frameworkElement, (int)args.OldValue);
                incrementalUpdater.CachePhaseElement(frameworkElement, (int)args.NewValue);
            }
        }

        private static IncrementalUpdater GetIncrementalUpdater(DependencyObject dependencyObject)
        {
            return (IncrementalUpdater)dependencyObject.GetValue(IncrementalUpdateBehavior.IncrementalUpdaterProperty);
        }

        private static void SetIncrementalUpdater(DependencyObject dependencyObject, IncrementalUpdater value)
        {
            dependencyObject.SetValue(IncrementalUpdateBehavior.IncrementalUpdaterProperty, value);
        }

        private static void OnIncrementalUpdaterChanged(DependencyObject sender, DependencyPropertyChangedEventArgs args)
        {
            if (args.OldValue != null)
            {
                IncrementalUpdater incrementalUpdater = (IncrementalUpdater)args.OldValue;
                incrementalUpdater.Detach();
            }
            if (args.NewValue != null)
            {
                IncrementalUpdater incrementalUpdater = (IncrementalUpdater)args.NewValue;
                incrementalUpdater.Attach(sender);
            }
        }

        private void OnAssociatedObjectLoaded(object sender, RoutedEventArgs e)
        {
            IncrementalUpdater incrementalUpdater = this.FindUpdater();

            if (incrementalUpdater != null)
            {
                incrementalUpdater.CachePhaseElement(this.AssociatedObject, this.Phase);
            }
        }

        private void OnAssociatedObjectUnloaded(object sender, RoutedEventArgs e)
        {
            if (this._updater != null)
            {
                this._updater.UncachePhaseElement(this.AssociatedObject, this.Phase);
            }

            this._updater = null;
        }

        private IncrementalUpdater FindUpdater()
        {
            if (this._updater != null)
            {
                return this._updater;
            }

            DependencyObject ancestor = this.AssociatedObject;
            while (ancestor != null)
            {
                DependencyObject parent = VisualTreeHelper.GetParent(ancestor);
                ListViewBase listView = parent as ListViewBase;

                if (listView != null)
                {
                    IncrementalUpdater currentUpdater = IncrementalUpdateBehavior.GetIncrementalUpdater(listView);
                    if (currentUpdater == null)
                    {
                        currentUpdater = new IncrementalUpdater();

                        IncrementalUpdateBehavior.SetIncrementalUpdater(listView, currentUpdater);
                    }
                    return currentUpdater;
                }
                ancestor = parent;
            }

            return null;
        }

        /// <summary>
        /// Called after the behavior is attached to the <see cref="Microsoft.Xaml.Interactivity.Behavior.AssociatedObject"/>.
        /// </summary>
        protected override void OnAttached()
        {
            base.OnAttached();

            this.AssociatedObject.Loaded += this.OnAssociatedObjectLoaded;
            this.AssociatedObject.Unloaded += this.OnAssociatedObjectUnloaded;
        }

        /// <summary>
        /// Called when the behavior is being detached from its <see cref="Microsoft.Xaml.Interactivity.Behavior.AssociatedObject"/>.
        /// </summary>
        protected override void OnDetaching()
        {
            base.OnDetaching();

            this.AssociatedObject.Loaded -= this.OnAssociatedObjectLoaded;
            this.AssociatedObject.Unloaded -= this.OnAssociatedObjectUnloaded;
            // no need to perform the work that Unloaded would have done - that's just housekeeping on the cache, which is now going away
        }

        private class IncrementalUpdater
        {
            private ListViewBase _associatedListViewBase = null;
            private Dictionary<UIElement, ElementCacheRecord> _elementCache = new Dictionary<UIElement, ElementCacheRecord>();

            private class PhasedElementRecord
            {
                private readonly FrameworkElement _frameworkElement;
                private object _localOpacity;
                private object _localDataContext;
                private bool _isFrozen;

                public PhasedElementRecord(FrameworkElement frameworkElement)
                {
                    this._frameworkElement = frameworkElement;
                }

                public FrameworkElement FrameworkElement { get { return this._frameworkElement; } }

                public void FreezeAndHide()
                {
                    if (this._isFrozen)
                    {
                        return;
                    }

                    this._isFrozen = true;
                    this._localOpacity = this._frameworkElement.ReadLocalValue(FrameworkElement.OpacityProperty);
                    this._localDataContext = this._frameworkElement.ReadLocalValue(FrameworkElement.DataContextProperty);
                    this._frameworkElement.Opacity = 0.0;
                    this._frameworkElement.DataContext = this._frameworkElement.DataContext;
                }

                public void ThawAndShow()
                {
                    if (!this._isFrozen)
                    {
                        return;
                    }

                    if (this._localOpacity != DependencyProperty.UnsetValue)
                    {
                        this._frameworkElement.SetValue(FrameworkElement.OpacityProperty, this._localOpacity);
                    }
                    else
                    {
                        this._frameworkElement.ClearValue(FrameworkElement.OpacityProperty);
                    }

                    if (this._localDataContext != DependencyProperty.UnsetValue)
                    {
                        this._frameworkElement.SetValue(FrameworkElement.DataContextProperty, this._localDataContext);
                    }
                    else
                    {
                        this._frameworkElement.ClearValue(FrameworkElement.DataContextProperty);
                    }

                    this._isFrozen = false;
                }
            }

            private class ElementCacheRecord
            {
                private List<int> _phases = new List<int>();
                private List<List<PhasedElementRecord>> _elementsByPhase = new List<List<PhasedElementRecord>>();

                public List<int> Phases
                {
                    get { return this._phases; }
                }

                public List<List<PhasedElementRecord>> ElementsByPhase
                {
                    get { return this._elementsByPhase; }
                }
            }

            private void OnContainerContentChanging(ListViewBase sender, ContainerContentChangingEventArgs e)
            {
                UIElement contentTemplateRoot = e.ItemContainer.ContentTemplateRoot;

                ElementCacheRecord elementCacheRecord;
                if (this._elementCache.TryGetValue(contentTemplateRoot, out elementCacheRecord))
                {
                    if (!e.InRecycleQueue)
                    {
                        foreach (List<PhasedElementRecord> phaseRecord in elementCacheRecord.ElementsByPhase)
                        {
                            foreach (PhasedElementRecord phasedElementRecord in phaseRecord)
                            {
                                phasedElementRecord.FreezeAndHide();
                            }
                        }

                        if (elementCacheRecord.Phases.Count > 0)
                        {
                            e.RegisterUpdateCallback((uint)elementCacheRecord.Phases[0], this.OnContainerContentChangingCallback);
                        }

                        // set the DataContext manually since we inhibit default operation by setting e.Handled=true
                        ((FrameworkElement)contentTemplateRoot).DataContext = e.Item;
                    }
                    else
                    {
                        // clear the DataContext manually since we inhibit default operation by setting e.Handled=true
                        contentTemplateRoot.ClearValue(FrameworkElement.DataContextProperty);

                        foreach (List<PhasedElementRecord> phaseRecord in elementCacheRecord.ElementsByPhase)
                        {
                            foreach (PhasedElementRecord phasedElementRecord in phaseRecord)
                            {
                                phasedElementRecord.ThawAndShow();
                            }
                        }
                    }
                }
            }

            private void OnContainerContentChangingCallback(ListViewBase sender, ContainerContentChangingEventArgs e)
            {
                UIElement contentTemplateRoot = e.ItemContainer.ContentTemplateRoot;

                ElementCacheRecord elementCacheRecord;
                if (this._elementCache.TryGetValue(contentTemplateRoot, out elementCacheRecord))
                {
                    int phaseIndex = elementCacheRecord.Phases.BinarySearch((int)e.Phase);

                    if (phaseIndex >= 0)
                    {
                        foreach (PhasedElementRecord phasedElementRecord in elementCacheRecord.ElementsByPhase[phaseIndex])
                        {
                            phasedElementRecord.ThawAndShow();
                        }

                        phaseIndex++;
                    }
                    else
                    {
                        // don't know why this phase was not found, but by BinarySearch rules, ~phaseIndex is the place
                        // where it would be inserted, thus the item there has the next higher number.
                        phaseIndex = ~phaseIndex;
                    }

                    if (phaseIndex < elementCacheRecord.Phases.Count)
                    {
                        e.RegisterUpdateCallback((uint)elementCacheRecord.Phases[phaseIndex], this.OnContainerContentChangingCallback);
                    }
                }
            }

            private static UIElement FindContentTemplateRoot(FrameworkElement phaseElement)
            {
                DependencyObject ancestor = phaseElement;
                while (ancestor != null)
                {
                    DependencyObject parent = VisualTreeHelper.GetParent(ancestor);
                    SelectorItem item = parent as SelectorItem;

                    if (item != null)
                    {
                        return item.ContentTemplateRoot;
                    }
                    ancestor = parent;
                }

                return null;
            }

            public void CachePhaseElement(FrameworkElement phaseElement, int phase)
            {
                if (phase < 0)
                {
                    throw new ArgumentOutOfRangeException(nameof(phase));
                }

                if (phase <= 0)
                {
                    return;
                }

                UIElement contentTemplateRoot = IncrementalUpdater.FindContentTemplateRoot(phaseElement);
                if (contentTemplateRoot != null)
                {
                    // get the cache for this element
                    ElementCacheRecord elementCacheRecord;
                    if (!this._elementCache.TryGetValue(contentTemplateRoot, out elementCacheRecord))
                    {
                        elementCacheRecord = new ElementCacheRecord();
                        this._elementCache.Add(contentTemplateRoot, elementCacheRecord);
                    }

                    // get the cache for this phase
                    int phaseIndex = elementCacheRecord.Phases.BinarySearch(phase);

                    if (phaseIndex < 0)
                    {
                        // not found - insert
                        phaseIndex = ~phaseIndex;
                        elementCacheRecord.Phases.Insert(phaseIndex, phase);
                        elementCacheRecord.ElementsByPhase.Insert(phaseIndex, new List<PhasedElementRecord>());
                    }

                    List<PhasedElementRecord> phasedElementRecords = elementCacheRecord.ElementsByPhase[phaseIndex];

                    // first see if the element is already there
                    for (int i = 0; i < phasedElementRecords.Count; i++)
                    {
                        if (phasedElementRecords[i].FrameworkElement == phaseElement)
                        {
                            return;
                        }
                    }

                    // insert the element
                    phasedElementRecords.Add(new PhasedElementRecord(phaseElement));
                }
            }

            public void UncachePhaseElement(FrameworkElement phaseElement, int phase)
            {
                if (phase <= 0)
                {
                    return;
                }

                UIElement contentTemplateRoot = IncrementalUpdater.FindContentTemplateRoot(phaseElement);
                if (contentTemplateRoot != null)
                {
                    // get the cache for this element
                    ElementCacheRecord elementCacheRecord;
                    if (this._elementCache.TryGetValue(contentTemplateRoot, out elementCacheRecord))
                    {
                        // get the cache for this phase
                        int phaseIndex = elementCacheRecord.Phases.BinarySearch(phase);

                        if (phaseIndex >= 0)
                        {
                            // remove the element: the linear search here is not spectacular but the list should be very short
                            List<PhasedElementRecord> phasedElementRecords = elementCacheRecord.ElementsByPhase[phaseIndex];

                            for (int i = 0; i < phasedElementRecords.Count; i++)
                            {
                                if (phasedElementRecords[i].FrameworkElement == phaseElement)
                                {
                                    phasedElementRecords[i].ThawAndShow();

                                    phasedElementRecords.RemoveAt(i);

                                    if (phasedElementRecords.Count == 0)
                                    {
                                        elementCacheRecord.Phases.RemoveAt(phaseIndex);
                                        elementCacheRecord.ElementsByPhase.RemoveAt(phaseIndex);
                                    }
                                }
                            }
                        }
                    }
                }
            }

            public void Attach(DependencyObject dependencyObject)
            {
                this._associatedListViewBase = dependencyObject as ListViewBase;

                if (this._associatedListViewBase != null)
                {
                    this._associatedListViewBase.ContainerContentChanging += this.OnContainerContentChanging;
                }
            }

            public void Detach()
            {
                if (this._associatedListViewBase != null)
                {
                    this._associatedListViewBase.ContainerContentChanging -= this.OnContainerContentChanging;
                }
                this._associatedListViewBase = null;
            }
        }
    }
}
